pandas/core/arrays/arrow/array.py

```diff
@@ -127,12 +127,26 @@ def floordiv_compat(
         left: pa.ChunkedArray | pa.Array | pa.Scalar,
         right: pa.ChunkedArray | pa.Array | pa.Scalar,
     ) -> pa.ChunkedArray:
-        # Ensure int // int -> int mirroring Python/Numpy behavior
-        # as pc.floor(pc.divide_checked(int, int)) -> float
-        converted_left = cast_for_truediv(left, right)
-        result = pc.floor(pc.divide(converted_left, right))
-        if pa.types.is_integer(left.type) and pa.types.is_integer(right.type):
+        divided = pc.divide(left, right)
+        if pa.types.is_integer(divided.type):
+            # GH 56676, avoid storing intermediate calculating in floating point type.
+            has_remainder = pc.not_equal(pc.multiply(divided, right), left)
+            result = pc.if_else(
+                pc.less(divided, 0),
+                # Avoid using subtract_checked which has significantly worse perf:
+                # https://github.com/apache/arrow/issues/37678, especially since
+                # integer overflow should be impossible here.
+                pc.if_else(has_remainder, pc.subtract(divided, 1), divided),
+                divided,
+            )
+            # Ensure compatibility with older versions of pandas where
+            # int8 // int64 returned int8 rather than int64.
```

Comment:
Is this something that can be changed? If so, the cast to left.type can be removed, result.type is guaranteed to already be integral.

Reply:
I actually don't think this old behavior is desirable, for example:

```
>>> a = pd.Series([-128], dtype="int8[pyarrow]")
>>> b = pd.Series([-1], dtype="int64[pyarrow]")
>>> a // b
```

With this cast, this operation fails with overflow error, because 128 can't fit in an int8. In numpy, it looks like this operation promotes to common type of int64:

```
>>> a = pd.Series([-128], dtype="int8")
>>> b = pd.Series([-1], dtype="int64")
>>> a // b
0    128
dtype: int64
>>>
```
So I think this cast should be removed.

Reply:
Looks like there's comments in the test about ArrowExtensionArray not promoting, so maybe this is the intended behavior? Restored the cast.